src/libs/data-utils.js

```diff
@@ -0,0 +1,121 @@
+import _ from 'lodash/fp'
+import filesize from 'filesize'
+import { Fragment } from 'react'
+import { code, div, h, pre } from 'react-hyperscript-helpers/lib/index'
+import Collapse from 'src/components/Collapse'
+import { buttonPrimary, link } from 'src/components/common'
+import { spinner } from 'src/components/icons'
+import Modal from 'src/components/Modal'
+import { TextCell } from 'src/components/table'
+import { Buckets, Martha } from 'src/libs/ajax'
+import * as Utils from 'src/libs/utils'
+import { Component } from 'src/libs/wrapped-components'
+
+
+const styles = {
+  cell: children => div({ style: { marginBottom: '0.5rem' } }, children),
+  label: text => div({}, text),
+  data: children => div({}, children)
+}
+
+const isFilePreviewable = name => /\.txt$/.test(name) || /\.[ct]sv$/.test(name) || /\.log/.test(name)
+const formatPriceInDollars = price => price < 0.01 ? '< $0.01' : `$${parseFloat(price).toFixed(2)}`
+const parseUri = uri => _.drop(1, /gs:[/][/]([^/]+)[/](.+)/.exec(uri))
+
+/**
+ * @param uri
+ * @param googleProject
+ */
+export class UriViewer extends Component {
+  constructor(props) {
+    super(props)
+
+    const { uri } = props
+
+    this.state = { uri: _.startsWith('gs://', uri) ? uri : undefined }
+  }
+
+  async getMetadata() {
+    const { googleProject } = this.props
+    const { uri } = this.state
```

Comment:
In the case of the uri being resolved by Martha, I don't think the `setState` is guaranteed to be visible by the time we get here. It might be safer to pass the value directly, and/or combine these methods.